File: bots/cpp/src/bot_api.cpp
// TigerBots C++ API client implementation (libcurl).
//
// See bot_api.hpp for the contract. All methods issue real HTTP calls via a
// libcurl easy handle. Fail-closed: any non-2xx response throws
// `BotApiException` carrying the backend's `error`/`message` text.

#include "bot_api.hpp"

#include <curl/curl.h>

#include <algorithm>
#include <cctype>
#include <cstddef>
#include <iomanip>
#include <sstream>

namespace tigerbots {

namespace {

class CurlGlobal {
public:
    CurlGlobal() { curl_global_init(CURL_GLOBAL_DEFAULT); }
    ~CurlGlobal() { curl_global_cleanup(); }
};

// RAII guard for libcurl's process-global init/cleanup. Created once on first
// use so the client can be compiled into libraries that don't control main().
CurlGlobal& curlGlobal() {
    static CurlGlobal g;
    return g;
}

struct WriteCtx {
    std::string buf;
};

size_t writeCb(char* ptr, size_t size, size_t nmemb, void* userdata) {
    auto* ctx = static_cast<WriteCtx*>(userdata);
    const size_t total = size * nmemb;
    ctx->buf.append(ptr, total);
    return total;
}

size_t headerCb(char* /*ptr*/, size_t size, size_t nmemb, void* /*userdata*/) {
    return size * nmemb;
}

} // namespace

BotApiClient::BotApiClient(std::string baseUrl)
    : base_url_(std::move(baseUrl)) {
    curlGlobal();
}

BotApiClient::~BotApiClient() = default;

std::string BotApiClient::encodePathSegment(const std::string& s) {
    std::ostringstream out;
    out << std::hex << std::uppercase << std::setfill('0');
    for (unsigned char c : s) {
        if (std::isalnum(c) || c == '-' || c == '_' || c == '.' || c == '~') {
            out << static_cast<char>(c);
        } else {
            out << '%' << std::setw(2) << static_cast<int>(c);
        }
    }
    return out.str();
}

nlohmann::json BotApiClient::request(const std::string& path,
                                     const std::string& method,
                                     const std::optional<nlohmann::json>& body,
                                     bool authenticated) {
    CURL* easy = curl_easy_init();
    if (!easy) {
        throw BotApiException(0, "curl_easy_init failed");
    }
    WriteCtx ctx;
    struct curl_slist* headers = nullptr;
    std::string bodyStr;
    try {
        headers = curl_slist_append(headers, "Accept: application/json");
        if (body.has_value()) {
            bodyStr = body->dump();
            headers = curl_slist_append(headers, "Content-Type: application/json; charset=utf-8");
        }
        if (authenticated) {
            if (!token_.has_value()) {
                curl_slist_free_all(headers);
                curl_easy_cleanup(easy);
                throw BotApiException(401, "not authenticated: no JWT token set");
            }
            const std::string auth = "Authorization: Bearer " + *token_;
            headers = curl_slist_append(headers, auth.c_str());
        }
        const std::string url = base_url_ + path;
        curl_easy_setopt(easy, CURLOPT_URL, url.c_str());
        curl_easy_setopt(easy, CURLOPT_CUSTOMREQUEST, method.c_str());
        curl_easy_setopt(easy, CURLOPT_HTTPHEADER, headers);
        curl_easy_setopt(easy, CURLOPT_TIMEOUT, 30L);
        curl_easy_setopt(easy, CURLOPT_CONNECTTIMEOUT, 15L);
        curl_easy_setopt(easy, CURLOPT_FOLLOWLOCATION, 0L);
        curl_easy_setopt(easy, CURLOPT_WRITEFUNCTION, writeCb);
        curl_easy_setopt(easy, CURLOPT_WRITEDATA, &ctx);
        curl_easy_setopt(easy, CURLOPT_HEADERFUNCTION, headerCb);
        curl_easy_setopt(easy, CURLOPT_NOSIGNAL, 1L);
        if (body.has_value()) {
            curl_easy_setopt(easy, CURLOPT_POSTFIELDS, bodyStr.c_str());
            curl_easy_setopt(easy, CURLOPT_POSTFIELDSIZE,
                             static_cast<long>(bodyStr.size()));
        }
        CURLcode rc = curl_easy_perform(easy);
        if (rc != CURLE_OK) {
            curl_slist_free_all(headers);
            curl_easy_cleanup(easy);
            throw BotApiException(0, std::string("curl: ") + curl_easy_strerror(rc));
        }
        long httpCode = 0;
        curl_easy_getinfo(easy, CURLINFO_RESPONSE_CODE, &httpCode);
        curl_slist_free_all(headers);
        curl_easy_cleanup(easy);
        if (httpCode < 200 || httpCode > 299) {
            std::string msg;
            try {
                auto j = nlohmann::json::parse(ctx.buf.empty() ? "{}" : ctx.buf);
                if (j.contains("error") && j["error"].is_string()) {
                    msg = j["error"].get<std::string>();
                } else if (j.contains("message") && j["message"].is_string()) {
                    msg = j["message"].get<std::string>();
                }
            } catch (...) {
                msg = ctx.buf;
            }
            if (msg.empty()) msg = "HTTP " + std::to_string(httpCode);
            throw BotApiException(static_cast<int>(httpCode), msg);
        }
        if (ctx.buf.empty()) return nlohmann::json{};
        return nlohmann::json::parse(ctx.buf);
    } catch (...) {
        if (headers) curl_slist_free_all(headers);
        curl_easy_cleanup(easy);
        throw;
    }
}

nlohmann::json BotApiClient::get(const std::string& path, bool authenticated) {
    return request(path, "GET", std::nullopt, authenticated);
}

nlohmann::json BotApiClient::del(const std::string& path) {
    return request(path, "DELETE", std::nullopt, true);
}

nlohmann::json BotApiClient::post(const std::string& path, const nlohmann::json& body) {
    return request(path, "POST", body, true);
}

nlohmann::json BotApiClient::put(const std::string& path, const nlohmann::json& body) {
    return request(path, "PUT", body, true);
}

nlohmann::json BotApiClient::postNoBody(const std::string& path) {
    return request(path, "POST", std::nullopt, true);
}

// -- Auth --

BotAuthResponse BotApiClient::registerUser(const std::string& username,
                                            const std::string& password,
                                            const std::optional<std::string>& email,
                                            const std::optional<std::string>& walletAddress,
                                            const std::optional<std::string>& role) {
    nlohmann::json body = {
        {"username", username},
        {"password", password},
    };
    if (email) body["email"] = *email;
    if (walletAddress) body["wallet_address"] = *walletAddress;
    if (role) body["role"] = *role;
    auto res = request("/auth/register", "POST", body, false);
    BotAuthResponse out{};
    if (res.contains("token") && res["token"].is_string()) {
        out.token = res["token"].get<std::string>();
        token_ = out.token;
    }
    out.userId = res.value("user_id", "");
    out.role = res.value("role", "");
    return out;
}

BotAuthResponse BotApiClient::login(const std::string& username, const std::string& password) {
    nlohmann::json body = {{"username", username}, {"password", password}};
    auto res = request("/auth/login", "POST", body, false);
    BotAuthResponse out{};
    if (res.contains("token") && res["token"].is_string()) {
        out.token = res["token"].get<std::string>();
        token_ = out.token;
    }
    out.userId = res.value("user_id", "");
    out.role = res.value("role", "");
    return out;
}

void BotApiClient::logout() {
    try {
        postNoBody("/auth/logout");
    } catch (...) {
        token_.reset();
        throw;
    }
    token_.reset();
}

// -- Health + public tiers --

nlohmann::json BotApiClient::health() { return get("/health", false); }
nlohmann::json BotApiClient::publicTiers() { return get("/public/tiers", false); }

// -- Bots CRUD + lifecycle --

nlohmann::json BotApiClient::listBots() { return get("/bots"); }
nlohmann::json BotApiClient::getBot(const std::string& id) {
    return get("/bots/" + encodePathSegment(id));
}
nlohmann::json BotApiClient::createBot(const std::string& name,
                                       const std::string& botType,
                                       const nlohmann::json& config,
                                       const std::optional<std::string>& exchange,
                                       const std::optional<std::string>& pair) {
    nlohmann::json body = {{"name", name}, {"bot_type", botType}, {"config", config}};
    if (exchange) body["exchange"] = *exchange;
    if (pair) body["pair"] = *pair;
    return post("/bots", body);
}
nlohmann::json BotApiClient::deleteBot(const std::string& id) {
    return del("/bots/" + encodePathSegment(id));
}
nlohmann::json BotApiClient::startBot(const std::string& id) {
    return postNoBody("/bots/" + encodePathSegment(id) + "/start");
}
nlohmann::json BotApiClient::stopBot(const std::string& id) {
    return postNoBody("/bots/" + encodePathSegment(id) + "/stop");
}
nlohmann::json BotApiClient::pauseBot(const std::string& id) {
    return postNoBody("/bots/" + encodePathSegment(id) + "/pause");
}
nlohmann::json BotApiClient::listBotExecutions(const std::string& id) {
    return get("/bots/" + encodePathSegment(id) + "/executions");
}
nlohmann::json BotApiClient::listBotLogs(const std::string& id) {
    return get("/bots/" + encodePathSegment(id) + "/logs");
}
nlohmann::json BotApiClient::listBotInstances() { return get("/bots/instances"); }
nlohmann::json BotApiClient::currentBotUser() { return get("/bots/me"); }

// -- Bot users --

nlohmann::json BotApiClient::listBotUsers() { return get("/bots/users"); }
nlohmann::json BotApiClient::createBotUser(const std::string& username,
                                           const std::string& password,
                                           const std::optional<std::string>& email,
                                           const std::optional<std::string>& walletAddress,
                                           const std::optional<std::string>& role) {
    nlohmann::json body = {{"username", username}, {"password", password}};
    if (email) body["email"] = *email;
    if (walletAddress) body["wallet_address"] = *walletAddress;
    if (role) body["role"] = *role;
    return post("/bots/users", body);
}
nlohmann::json BotApiClient::deleteBotUser(const std::string& id) {
    return del("/bots/users/" + encodePathSegment(id));
}
nlohmann::json BotApiClient::listBotTransactions() { return get("/bots/transactions"); }

// -- Subscriptions --

nlohmann::json BotApiClient::getSubscription() { return get("/subscription"); }
nlohmann::json BotApiClient::createSubscription(const std::string& tier,
                                                const std::optional<std::string>& expiresIn) {
    nlohmann::json body = {{"tier", tier}};
    if (expiresIn) body["expires_in"] = *expiresIn;
    return post("/subscription", body);
}

// -- Fees --

nlohmann::json BotApiClient::getFeeConfigs() { return get("/fees"); }
nlohmann::json BotApiClient::updateFeeConfig(const std::string& id,
                                             const std::optional<std::string>& name,
                                             const std::optional<std::string>& percentage,
                                             const std::optional<bool>& enabled) {
    nlohmann::json body = {{"id", id}};
    if (name) body["name"] = *name;
    if (percentage) body["percentage"] = *percentage;
    if (enabled) body["enabled"] = *enabled;
    return put("/fees", body);
}

// -- CEX connectors --

nlohmann::json BotApiClient::listCEX() { return get("/cex"); }
nlohmann::json BotApiClient::addCEX(const std::string& name, const nlohmann::json& config) {
    return post("/cex", {{"name", name}, {"config", config}});
}
nlohmann::json BotApiClient::removeCEX(const std::string& id) {
    return del("/cex/" + encodePathSegment(id));
}

// -- DEX connectors --

nlohmann::json BotApiClient::listDEX() { return get("/dex"); }
nlohmann::json BotApiClient::addDEX(const std::string& name, const nlohmann::json& config) {
    return post("/dex", {{"name", name}, {"config", config}});
}
nlohmann::json BotApiClient::removeDEX(const std::string& id) {
    return del("/dex/" + encodePathSegment(id));
}

// -- API keys --

nlohmann::json BotApiClient::listAPIKeys() { return get("/keys"); }
nlohmann::json BotApiClient::createAPIKey(const std::string& exchange, const std::string& apiKey) {
    return post("/keys", {{"exchange", exchange}, {"api_key", apiKey}});
}
nlohmann::json BotApiClient::deleteAPIKey(const std::string& id) {
    return del("/keys/" + encodePathSegment(id));
}

// -- Admin --

nlohmann::json BotApiClient::adminListUsers() { return get("/admin/users"); }
nlohmann::json BotApiClient::adminUserStatus(const std::string& id, bool active) {
    nlohmann::json body = {{"id", id}, {"is_active", active}};
    return put("/admin/users/" + encodePathSegment(id) + "/status", body);
}
nlohmann::json BotApiClient::adminStats() { return get("/admin/stats"); }
nlohmann::json BotApiClient::adminGetFeeAddresses() { return get("/admin/fee-addresses"); }
nlohmann::json BotApiClient::adminSetFeeAddress(const std::string& label,
                                                long long chainId,
                                                const std::string& address) {
    return post("/admin/fee-addresses",
                {{"label", label}, {"chain_id", chainId}, {"address", address}});
}
nlohmann::json BotApiClient::adminDeleteFeeAddress(const std::string& id) {
    return del("/admin/fee-addresses/" + encodePathSegment(id));
}
nlohmann::json BotApiClient::adminBotStatus(const std::string& id, const std::string& status) {
    return post("/admin/bots/" + encodePathSegment(id) + "/status",
                {{"id", id}, {"status", status}});
}

} // namespace tigerbots

File: bots/rust/src/bot_api.rs
//! TigerBots Rust client.
//!
//! Async reqwest client delegating to the standalone Bots backend
//! (`mm_bot_platform/bot_api`, port 8471, path prefix `/api/v1`). JWT bearer
//! auth; the token is held in-process (`Arc<RwLock<Option<String>>>`) — Rust
//! clients are server/CLI and have no keychain dependency by default; callers
//! that need persistence can wrap `set_token`/`token`.
//!
//! Every method issues a real reqwest call against the backend — no stubs,
//! fakes, or mock data. On any non-2xx response the method returns an `Err`
//! carrying the backend's `error` message (fail-closed); it never returns
//! fabricated data.
//!
//! Method set mirrors `bots/web/src/services/api.ts`:
//!   auth: `login`, `register`, `logout`
//!   bots: `list_bots`, `get_bot`, `create_bot`, `delete_bot`, `start_bot`,
//!         `stop_bot`, `pause_bot`, `list_bot_executions`, `list_bot_logs`,
//!         `list_bot_instances`, `current_bot_user`
//!   users: `list_bot_users`, `create_bot_user`, `delete_bot_user`,
//!         `list_bot_transactions`
//!   subscriptions: `get_subscription`, `create_subscription`
//!   fees: `get_fee_configs`, `update_fee_config`
//!   cex: `list_cex`, `add_cex`, `remove_cex`
//!   dex: `list_dex`, `add_dex`, `remove_dex`
//!   keys: `list_api_keys`, `create_api_key`, `delete_api_key`
//!   admin: `admin_list_users`, `admin_user_status`, `admin_stats`,
//!         `admin_get_fee_addresses`, `admin_set_fee_address`,
//!         `admin_delete_fee_address`, `admin_bot_status`
//!   public: `public_tiers`, `health`

use reqwest::{Client, Method};
use serde_json::Value;
use std::sync::{Arc, RwLock};
use std::time::Duration;

pub const BOTS_API_DEFAULT_URL: &str = "http://localhost:8471/api/v1";

/// Error returned by every [`BotsClient`] method. `Http` carries the status
/// code and the backend's `error`/`message` text (or raw body). Fail-closed:
/// the client never fabricates a response on failure.
#[derive(Debug)]
pub enum BotError {
    Network(String),
    Decode(String),
    Http { status: u16, message: String },
    Unauthorized,
}

impl std::fmt::Display for BotError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            BotError::Network(m) => write!(f, "network: {m}"),
            BotError::Decode(m) => write!(f, "decode: {m}"),
            BotError::Http { status, message } => write!(f, "http {status}: {message}"),
            BotError::Unauthorized => write!(f, "not authenticated: no JWT token set"),
        }
    }
}

impl std::error::Error for BotError {}

/// Pooled async HTTP client for the standalone Bots backend.
pub struct BotsClient {
    http: Client,
    base_url: String,
    token: Arc<RwLock<Option<String>>>,
}

impl BotsClient {
    pub fn new(base_url: impl Into<String>, token: Option<String>) -> Self {
        let http = Client::builder()
            .timeout(Duration::from_secs(30))
            .pool_max_idle_per_host(16)
            .tcp_nodelay(true)
            .build()
            .expect("reqwest client build");
        Self {
            http,
            base_url: base_url.into(),
            token: Arc::new(RwLock::new(token)),
        }
    }

    /// Convenience constructor using [`BOTS_API_DEFAULT_URL`] and no token.
    pub fn default() -> Self {
        Self::new(BOTS_API_DEFAULT_URL, None)
    }

    pub fn set_token(&self, token: Option<String>) {
        *self.token.write().unwrap() = token;
    }

    pub fn token(&self) -> Option<String> {
        self.token.read().unwrap().clone()
    }

    pub fn base_url(&self) -> &str {
        &self.base_url
    }

    // -- internal request helpers --

    fn url(&self, path: &str) -> String {
        format!("{}{}", self.base_url, path)
    }

    async fn request(
        &self,
        method: Method,
        path: &str,
        body: Option<Value>,
        authenticated: bool,
    ) -> Result<Value, BotError> {
        let url = self.url(path);
        let mut req = self.http.request(method, &url);
        if authenticated {
            if let Some(t) = self.token() {
                req = req.bearer_auth(t);
            } else {
                return Err(BotError::Unauthorized);
            }
        }
        if let Some(b) = body {
            req = req.json(&b);
        }
        let resp = req.send().await.map_err(|e| BotError::Network(e.to_string()))?;
        let status = resp.status();
        let bytes = resp
            .bytes()
            .await
            .map_err(|e| BotError::Network(e.to_string()))?;
        let text = String::from_utf8_lossy(&bytes);
        let value: Value = if text.trim().is_empty() {
            Value::Null
        } else {
            serde_json::from_str(&text).map_err(|e| BotError::Decode(e.to_string()))?
        };
        if !status.is_success() {
            let msg = value
                .get("error")
                .or_else(|| value.get("message"))
                .and_then(|v| v.as_str())
                .map(|s| s.to_string())
                .unwrap_or_else(|| {
                    if text.trim().is_empty() {
                        format!("HTTP {}", status.as_u16())
                    } else {
                        text.to_string()
                    }
                });
            return Err(BotError::Http {
                status: status.as_u16(),
                message: msg,
            });
        }
        Ok(value)
    }

    async fn get(&self, path: &str, authenticated: bool) -> Result<Value, BotError> {
        self.request(Method::GET, path, None, authenticated).await
    }

    async fn send_body(
        &self,
        method: Method,
        path: &str,
        body: Value,
    ) -> Result<Value, BotError> {
        self.request(method, path, Some(body), true).await
    }

    async fn post(&self, path: &str, body: Value) -> Result<Value, BotError> {
        self.send_body(Method::POST, path, body).await
    }

    async fn put(&self, path: &str, body: Value) -> Result<Value, BotError> {
        self.send_body(Method::PUT, path, body).await
    }

    async fn delete(&self, path: &str) -> Result<Value, BotError> {
        self.request(Method::DELETE, path, None, true).await
    }

    async fn post_no_body(&self, path: &str) -> Result<Value, BotError> {
        self.request(Method::POST, path, None, true).await
    }

    // -- Auth --

    pub async fn register(
        &self,
        username: &str,
        password: &str,
        email: Option<&str>,
        wallet_address: Option<&str>,
        role: Option<&str>,
    ) -> Result<Value, BotError> {
        let mut payload = serde_json::json!({ "username": username, "password": password });
        if let Some(e) = email {
            payload["email"] = Value::String(e.to_string());
        }
        if let Some(w) = wallet_address {
            payload["wallet_address"] = Value::String(w.to_string());
        }
        if let Some(r) = role {
            payload["role"] = Value::String(r.to_string());
        }
        let res = self
            .request(Method::POST, "/auth/register", Some(payload), false)
            .await?;
        if let Some(t) = res.get("token").and_then(|v| v.as_str()) {
            self.set_token(Some(t.to_string()));
        }
        Ok(res)
    }

    pub async fn login(&self, username: &str, password: &str) -> Result<Value, BotError> {
        let payload = serde_json::json!({ "username": username, "password": password });
        let res = self
            .request(Method::POST, "/auth/login", Some(payload), false)
            .await?;
        if let Some(t) = res.get("token").and_then(|v| v.as_str()) {
            self.set_token(Some(t.to_string()));
        }
        Ok(res)
    }

    pub async fn logout(&self) -> Result<(), BotError> {
        let res = self.post_no_body("/auth/logout").await;
        self.set_token(None);
        match res {
            Ok(_) => Ok(()),
            Err(e) => Err(e),
        }
    }

    // -- Health + public tiers --

    pub async fn health(&self) -> Result<Value, BotError> {
        self.get("/health", false).await
    }

    pub async fn public_tiers(&self) -> Result<Value, BotError> {
        self.get("/public/tiers", false).await
    }

    // -- Bots CRUD + lifecycle --

    pub async fn list_bots(&self) -> Result<Value, BotError> {
        self.get("/bots", true).await
    }

    pub async fn get_bot(&self, id: &str) -> Result<Value, BotError> {
        self.get(&format!("/bots/{}", encode(id)), true).await
    }

    pub async fn create_bot(
        &self,
        name: &str,
        bot_type: &str,
        config: Option<Value>,
        exchange: Option<&str>,
        pair: Option<&str>,
    ) -> Result<Value, BotError> {
        let mut payload = serde_json::json!({
            "name": name,
            "bot_type": bot_type,
            "config": config.unwrap_or_else(|| Value::Object(Default::default())),
        });
        if let Some(e) = exchange {
            payload["exchange"] = Value::String(e.to_string());
        }
        if let Some(p) = pair {
            payload["pair"] = Value::String(p.to_string());
        }
        self.post("/bots", payload).await
    }

    pub async fn delete_bot(&self, id: &str) -> Result<Value, BotError> {
        self.delete(&format!("/bots/{}", encode(id))).await
    }

    pub async fn start_bot(&self, id: &str) -> Result<Value, BotError> {
        self.post_no_body(&format!("/bots/{}/start", encode(id))).await
    }

    pub async fn stop_bot(&self, id: &str) -> Result<Value, BotError> {
        self.post_no_body(&format!("/bots/{}/stop", encode(id))).await
    }

    pub async fn pause_bot(&self, id: &str) -> Result<Value, BotError> {
        self.post_no_body(&format!("/bots/{}/pause", encode(id))).await
    }

    pub async fn list_bot_executions(&self, id: &str) -> Result<Value, BotError> {
        self.get(&format!("/bots/{}/executions", encode(id)), true)
            .await
    }

    pub async fn list_bot_logs(&self, id: &str) -> Result<Value, BotError> {
        self.get(&format!("/bots/{}/logs", encode(id)), true).await
    }

    pub async fn list_bot_instances(&self) -> Result<Value, BotError> {
        self.get("/bots/instances", true).await
    }

    pub async fn current_bot_user(&self) -> Result<Value, BotError> {
        self.get("/bots/me", true).await
    }

    // -- Bot users --

    pub async fn list_bot_users(&self) -> Result<Value, BotError> {
        self.get("/bots/users", true).await
    }

    pub async fn create_bot_user(
        &self,
        username: &str,
        password: &str,
        email: Option<&str>,
        wallet_address: Option<&str>,
        role: Option<&str>,
    ) -> Result<Value, BotError> {
        let mut payload = serde_json::json!({ "username": username, "password": password });
        if let Some(e) = email {
            payload["email"] = Value::String(e.to_string());
        }
        if let Some(w) = wallet_address {
            payload["wallet_address"] = Value::String(w.to_string());
        }
        if let Some(r) = role {
            payload["role"] = Value::String(r.to_string());
        }
        self.post("/bots/users", payload).await
    }

    pub async fn delete_bot_user(&self, id: &str) -> Result<Value, BotError> {
        self.delete(&format!("/bots/users/{}", encode(id))).await
    }

    pub async fn list_bot_transactions(&self) -> Result<Value, BotError> {
        self.get("/bots/transactions", true).await
    }

    // -- Subscriptions --

    pub async fn get_subscription(&self) -> Result<Value, BotError> {
        self.get("/subscription", true).await
    }

    pub async fn create_subscription(
        &self,
        tier: &str,
        expires_in: Option<&str>,
    ) -> Result<Value, BotError> {
        let mut payload = serde_json::json!({ "tier": tier });
        if let Some(e) = expires_in {
            payload["expires_in"] = Value::String(e.to_string());
        }
        self.post("/subscription", payload).await
    }

    // -- Fees --

    pub async fn get_fee_configs(&self) -> Result<Value, BotError> {
        self.get("/fees", true).await
    }

    pub async fn update_fee_config(
        &self,
        id: &str,
        name: Option<&str>,
        percentage: Option<&str>,
        enabled: Option<bool>,
    ) -> Result<Value, BotError> {
        let mut payload = serde_json::json!({ "id": id });
        if let Some(n) = name {
            payload["name"] = Value::String(n.to_string());
        }
        if let Some(p) = percentage {
            payload["percentage"] = Value::String(p.to_string());
        }
        if let Some(e) = enabled {
            payload["enabled"] = Value::Bool(e);
        }
        self.put("/fees", payload).await
    }

    // -- CEX connectors --

    pub async fn list_cex(&self) -> Result<Value, BotError> {
        self.get("/cex", true).await
    }

    pub async fn add_cex(&self, name: &str, config: Value) -> Result<Value, BotError> {
        self.post("/cex", serde_json::json!({ "name": name, "config": config }))
            .await
    }

    pub async fn remove_cex(&self, id: &str) -> Result<Value, BotError> {
        self.delete(&format!("/cex/{}", encode(id))).await
    }

    // -- DEX connectors --

    pub async fn list_dex(&self) -> Result<Value, BotError> {
        self.get("/dex", true).await
    }

    pub async fn add_dex(&self, name: &str, config: Value) -> Result<Value, BotError> {
        self.post("/dex", serde_json::json!({ "name": name, "config": config }))
            .await
    }

    pub async fn remove_dex(&self, id: &str) -> Result<Value, BotError> {
        self.delete(&format!("/dex/{}", encode(id))).await
    }

    // -- API keys --

    pub async fn list_api_keys(&self) -> Result<Value, BotError> {
        self.get("/keys", true).await
    }

    pub async fn create_api_key(&self, exchange: &str, api_key: &str) -> Result<Value, BotError> {
        self.post(
            "/keys",
            serde_json::json!({ "exchange": exchange, "api_key": api_key }),
        )
        .await
    }

    pub async fn delete_api_key(&self, id: &str) -> Result<Value, BotError> {
        self.delete(&format!("/keys/{}", encode(id))).await
    }

    // -- Admin (super-admin / finance-admin only) --

    pub async fn admin_list_users(&self) -> Result<Value, BotError> {
        self.get("/admin/users", true).await
    }

    pub async fn admin_user_status(&self, id: &str, active: bool) -> Result<Value, BotError> {
        self.put(
            &format!("/admin/users/{}/status", encode(id)),
            serde_json::json!({ "id": id, "is_active": active }),
        )
        .await
    }

    pub async fn admin_stats(&self) -> Result<Value, BotError> {
        self.get("/admin/stats", true).await
    }

    pub async fn admin_get_fee_addresses(&self) -> Result<Value, BotError> {
        self.get("/admin/fee-addresses", true).await
    }

    pub async fn admin_set_fee_address(
        &self,
        label: &str,
        chain_id: i64,
        address: &str,
    ) -> Result<Value, BotError> {
        self.post(
            "/admin/fee-addresses",
            serde_json::json!({ "label": label, "chain_id": chain_id, "address": address }),
        )
        .await
    }

    pub async fn admin_delete_fee_address(&self, id: &str) -> Result<Value, BotError> {
        self.delete(&format!("/admin/fee-addresses/{}", encode(id))).await
    }

    pub async fn admin_bot_status(&self, id: &str, status: &str) -> Result<Value, BotError> {
        self.post(
            &format!("/admin/bots/{}/status", encode(id)),
            serde_json::json!({ "id": id, "status": status }),
        )
        .await
    }
}

/// Percent-encode a path segment (ids may contain characters unsafe in a URL).
fn encode(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for b in s.as_bytes() {
        let c = *b as char;
        if c.is_ascii_alphanumeric() || c == '-' || c == '_' || c == '.' || c == '~' {
            out.push(c);
        } else {
            out.push_str(&format!("%{:02X}", b));
        }
    }
    out
}
